### PR TITLE
Add Entry General: Middleware

### PR DESCRIPTION
**Pull request by denisCabrera**
### Description

- Added brief description of middleware
- Added examples of different types of middleware

### Type of Change

- Adding a new entry

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.

### CODE REVIEW

**Comment by denisCabrera** [comment]
@Dusch4593 Hi, this is a draft. I just wanted to get some feedback. I was thinking about expanding a little more on the descriptions for the different types of middleware. Still getting a hang of making new pull requests. For some reason it keeps adding old files to this PR. Any thoughts ? Thank you 😃

**Comment by denisCabrera** [comment]
@yangc95 I already deleted the extra file. But it was adding the picture.md file I created a while back. That is what I was referring to. Even though I am creating a new branch a file from a previous PR seemed to be still be lingering. I am now reading more into it I am sure its an operator error from my part lol

**Comment by KTom101** [comment]
@denisCabrera I think this is off to a great start. Are you thinking of adding more types of middleware?  
In concern to the pull request, do you know if your fork is up to date? Maybe your local repo is behind in commits and it’s not seeing your old file as already being merged?

**Comment by denisCabrera** [comment]
@KTom101 I was thinking of just keeping those. And that might be it, I will make sure it is updated.

**Comment by KTom101** [comment]
@denisCabrera You can take it off draft whenever you're ready.

**Comment by denisCabrera** [comment]
@yangc95 @KTom101 @Dusch4593  This pull request is ready for review.

File: content/general/concepts/middleware/middleware.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+Title: 'Middleware'
+Description: 'Middleware is software that connects applications together. Middleware connects data, APIs, software tools and other applications in other to bring a complete application to the end user.'
+Subjects:
+  - 'Computer Science'
+  - 'Web Development'
+Tags:
+  - 'APIs'
+  - 'Database'
+  - 'Middleware'
+CatalogContent: # Please use course/path landing page slugs, rather than linking to individual content items. If listing multiple items, please put the most relevant one first
```
Comment (Dusch4593) [inline content/general/concepts/middleware/middleware.md:11]:
```suggestion
CatalogContent:
```

Those comments are just meant for templating purposes.

File: content/general/concepts/middleware/middleware.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  - 'learn-html'
```
Comment (Dusch4593) [inline content/general/concepts/middleware/middleware.md:12]: Is there another `learn-` slug that relates to middleware more closely?

File: content/general/concepts/middleware/middleware.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
+---
+Title: 'Middleware'
+Description: 'Middleware is software that connects applications together. Middleware connects data, APIs, software tools and other applications in other to bring a complete application to the end user.'
```
Comment (Dusch4593) [inline content/general/concepts/middleware/middleware.md:3]:
```suggestion
Description: 'Middleware is software that connects data, APIs, software tools, and other applications in order to bring a complete application to the end user.'
```

File: content/general/concepts/middleware/middleware.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+Tags:
+  - 'APIs'
+  - 'Database'
+  - 'Middleware'
+CatalogContent: # Please use course/path landing page slugs, rather than linking to individual content items. If listing multiple items, please put the most relevant one first
+  - 'learn-html'
+  - 'paths/back-end-engineer-career-path'
+---
+
+Middleware is software that connects applications together. The term _middleware_ comes from the fact that this software sits between the front-end or client side of an application and the back-end where the resources are being requested. Middleware connects or "glues" together data, APIs, software tools and other applications in order to bring a complete or bundled application to the end user. There are many types of middleware software depending on the use cases. Some of the most commonly used middleware is listed below.
```
Comment (Dusch4593) [inline content/general/concepts/middleware/middleware.md:16]:
```suggestion
Middleware is software that connects applications together. The term "middleware" comes from the fact that this software sits between the front-end or client side of an application and the back-end where the resources are being requested. Middleware connects or "glues" together data, APIs, software tools and other applications in order to bring a complete or bundled application to the end user. There are many types of middleware software depending on the use cases. Some of the most commonly used middleware is listed below.
```

Most forms of emphasis are generally advised against. ☹️

File: content/general/concepts/middleware/middleware.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+## Types of Middleware
+
+### Message-oriented middleware
```
Comment (Dusch4593) [inline content/general/concepts/middleware/middleware.md:20]:
```suggestion
### Message Oriented Middleware
```
For headers, let's use strict "Title Case" 😄

File: content/general/concepts/middleware/middleware.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+This type of middleware allows applications to exchange messages and communicate between eachother.
```
Comment (Dusch4593) [inline content/general/concepts/middleware/middleware.md:22]:
```suggestion
A message oriented middleware, or MOM, allows applications to exchange messages and communicate between one another.
```

Could include a short list of prominent examples for each of these middleware types?

File: content/general/concepts/middleware/middleware.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+### Database middleware
+
+This type of middleware gives access to databases and provides points of interaction between a variety of databases.
+
+### Application Programming Interface (API) Middleware
```
Comment (Dusch4593) [inline content/general/concepts/middleware/middleware.md:28]:
```suggestion
### Application Programming Interface
```

**Review by Dusch4593 — CHANGES_REQUESTED** [review]
@denisCabrera Finished with leaving comments/suggestions for this review. I agree with @KTom101, this entry is off to a great start! 👍 Feel free to tag one of us when you are finished with pushing new changes.

P.S. I thought [this](https://www.ietf.org/rfc/rfc2768.txt) would be worth sharing with everyone since we are working with middleware stuff. It is from a Feb. 2000 workshop report conducted by the Internet Engineering Task Force (IETF) re: middleware, what it is, what it is used for, and why. Fun read on pp. 4! (<kbd>Ctrl</kbd> + <kbd>F</kbd> + type "[Page 4]") Though a bit older, this page's description of the contextual nature of the definition of middleware seems like it still holds up today?

**Comment by denisCabrera** [comment]
@Dusch4593 Interesting read. Thank you for sharing. This report makes a great point about how the definition of middleware could be somewhat subjective. I took on this issue as a learning challenge and I quickly learned middleware is much more than I initially thought. 
@yangc95 @KTom101 I have made the suggested changes. If you need me to update anything else feel free to let me know. Hope everyone is having a great weekend! 😎

File: content/general/concepts/middleware/middleware.md (hunk 4 above)
Comment (yangc95) [reply content/general/concepts/middleware/middleware.md]: Oh, yeah ☹️ no italics for future reference

File: content/general/concepts/middleware/middleware.md (hunk 5 above)
Comment (yangc95) [reply content/general/concepts/middleware/middleware.md]: AP Style* Title Cased

File: content/general/concepts/middleware/middleware.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  - 'Database'
+  - 'Middleware'
+CatalogContent:
+  - 'learn-sql'
+  - 'paths/back-end-engineer-career-path'
+---
+
+Middleware is software that connects applications together. The term "middleware" comes from the fact that this software sits between the front-end, or client-side, of an application and the back-end, or server-side, where the resources are being requested.
+
+Data, APIs, software tools, and other applications are connected by middleware in order to bring a complete or bundled application to the end user. There are many types of middleware software depending on the use cases. Some of the most commonly used middleware is listed below.
```
Comment (yangc95) [inline content/general/concepts/middleware/middleware.md:18]: Thinking the last two sentences here can be shifted to after the "Types of Middleware" heading and before "Message Oriented Middleware"

File: content/general/concepts/middleware/middleware.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+### Database middleware
+
+This type of middleware gives access to databases and provides points of interaction between a variety of databases. This middleware improves efficiency however it can be language specific which can create limitations.
+
+#### Examples
+
+- ODBC
+
+Open Database Connectivity is an API used for accessing database management systems.
```
Comment (yangc95) [inline content/general/concepts/middleware/middleware.md:48]: Can we try placing these in a toggle. When we click on the toggle, it should expand to show the additional info on the middleware.

```
<details>
  <summary>ODBC</summary>
  Open Database Connectivity is an API used for accessing database management systems.
</details>
```

**Review by yangc95 — COMMENTED** [review]
Hi @denisCabrera, really like how everything is worded clearly and meaningfully. I made a few comments/suggestions mostly on the entry's structure.

File: content/general/concepts/middleware/middleware.md (hunk 9 above)
Comment (denisCabrera) [reply content/general/concepts/middleware/middleware.md:48]: > Can we try placing these in a toggle. When we click on the toggle, it should expand to show the additional info on the middleware.
> 
> ```
> <details>
>   <summary>ODBC</summary>
>   Open Database Connectivity is an API used for accessing database management systems.
> </details>
> ```

Should I do it for all the examples ? or just this one ?

**Comment by denisCabrera** [comment]
@yangc95 Hi, I made the suggested changes. I really like that you suggested the toggle option for all the examples. Thank you! 👍

**Review by yangc95 — APPROVED** [review]
LGTM 👍 what do you think 🧠 @Dusch4593 !

File: content/general/concepts/middleware/middleware.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+### API Middleware
+
+This type of middleware provides developers with the tools to create and manage APIs used in applications. This middleware offers scalability but connections can be slow when dealing with different servers.
+
+#### Examples
+
+<details>
+  <summary>API management platforms</summary>
+API management platforms allow developers to access, control, and analyze APIs.
```
Comment (yangc95) [inline content/general/concepts/middleware/middleware.md:67]:
```suggestion
  API management platforms allow developers to access, control, and analyze APIs.
```

File: content/general/concepts/middleware/middleware.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+</details>
+
+<details>
+  <summary>API gateways</summary>
+API gateways provide services related to user authentication and rate litmiting.
```
Comment (yangc95) [inline content/general/concepts/middleware/middleware.md:72]:
```suggestion
  API gateways provide services related to user authentication and rate litmiting.
```

File: content/general/concepts/middleware/middleware.md [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+#### Examples
+
+<details>
+  <summary>RabbitMQ</summary>
+RabbitMQ is a messaging broker platform used by applications to send and recieve messages.
+</details>
+
+<details>
+  <summary>IBM MQ</summary>
+IBM MQ is another example of message oriented middleware. Some of the core components include messages, message queues and queue manager.
```
Comment (Dusch4593) [inline content/general/concepts/middleware/middleware.md:35]:
```suggestion
  <summary>IBM MQ</summary>
  IBM MQ is another example of message oriented middleware. Some of the core components include messages, message queues and queue manager.
```

File: content/general/concepts/middleware/middleware.md [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+</details>
+
+<details>
+  <summary>JDBC</summary>
+Java Database Connectivity is an API that utilizes the Java programming language to access databases.
+</details>
+
+### API Middleware
+
+This type of middleware provides developers with the tools to create and manage APIs used in applications. This middleware offers scalability but connections can be slow when dealing with different servers.
```
Comment (Dusch4593) [inline content/general/concepts/middleware/middleware.md:61]:
```suggestion
This type of middleware provides developers with the tools to create and manage APIs used in applications. While it offers scalability, connections can be slow when dealing with different servers.
```

**Comment by Dusch4593** [comment]
Hey @denisCabrera I left a few additional comments/suggestions for you to check out when you get a sec 😄

**Comment by denisCabrera** [comment]
@Dusch4593 I finished all the suggestions. Thank you for catching those details that make the entry so much cleaner. 😄

**Review by Dusch4593 — APPROVED** [review]
> @Dusch4593 I finished all the suggestions. Thank you for catching those details that make the entry so much cleaner. 😄

👍

**Comment by yangc95** [comment]
Any last minutes on this one 😬 ? @Dusch4593 @KTom101 @SSwiniarski

File: content/general/concepts/middleware/middleware.md [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+CatalogContent:
+  - 'learn-sql'
+  - 'paths/back-end-engineer-career-path'
+---
+
+Middleware is software that connects applications together. The term "middleware" comes from the fact that this software sits between the front-end, or client-side, of an application and the back-end, or server-side, where the resources are being requested.
+
+## Types of Middleware
+
+Data, APIs, software tools, and other applications are connected by middleware in order to bring a complete or bundled application to the end user. There are many types of middleware software depending on the use cases. Some of the most commonly used middleware is listed below.
```
Comment (KTom101) [inline content/general/concepts/middleware/middleware.md:20]:
```suggestion
Data, APIs, software tools, and other applications are connected by middleware in order to bring a complete or bundled application to the end user. There are many types of middleware software depending on the use cases. Some of the most commonly used middleware are listed below.
```

File: content/general/concepts/middleware/middleware.md [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+### Message Oriented Middleware
+
+A message oriented middleware, or MOM, allows applications to exchange messages and communicate between one another. This middleware offers ease of use but can be costly at larger scales.
```
Comment (KTom101) [inline content/general/concepts/middleware/middleware.md:24]:
```suggestion
A message oriented middleware, or MOM, allows applications to exchange messages and communicate with one another. This middleware offers ease of use but can be costly at larger scales.
```

File: content/general/concepts/middleware/middleware.md [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+#### Examples
+
+<details>
+  <summary>API management platforms</summary>
+  API management platforms allow developers to access, control, and analyze APIs.
+</details>
+
+<details>
+  <summary>API gateways</summary>
+  API gateways provide services related to user authentication and rate litmiting.
```
Comment (KTom101) [inline content/general/concepts/middleware/middleware.md:72]:
```suggestion
  API gateways provide services related to user authentication and rate limiting.
```

**Review by KTom101 — COMMENTED** [review]
Very minor changes.

**Comment by denisCabrera** [comment]
@KTom101 All changes are complete. They were minor, but good catch on the grammar, thank you ! 😄

**Comment by Dusch4593** [comment]
Thank you, @denisCabrera, for another solid Docs contribution! 🙏 🎉 🎊 

When the entry goes live, you can check it out here: https://www.codecademy.com/resources/docs/general/middleware

**Comment by denisCabrera** [comment]
@yangc95 @KTom101 @Dusch4593 Thank you all for the great feedback. I am grateful for the opportunity 🙏

**Pull request merged**